src/roadveh_cmd.cpp

```diff
@@ -620,21 +620,24 @@ static Vehicle *EnumCheckRoadVehClose(Vehicle *v, void *data)
 	short x_diff = v->x_pos - rvf->x;
 	short y_diff = v->y_pos - rvf->y;
 
-	if (v->type == VEH_ROAD &&
-			!v->IsInDepot() &&
-			abs(v->z_pos - rvf->veh->z_pos) < 6 &&
-			v->direction == rvf->dir &&
-			rvf->veh->First() != v->First() &&
-			(dist_x[v->direction] >= 0 || (x_diff > dist_x[v->direction] && x_diff <= 0)) &&
-			(dist_x[v->direction] <= 0 || (x_diff < dist_x[v->direction] && x_diff >= 0)) &&
-			(dist_y[v->direction] >= 0 || (y_diff > dist_y[v->direction] && y_diff <= 0)) &&
-			(dist_y[v->direction] <= 0 || (y_diff < dist_y[v->direction] && y_diff >= 0))) {
-		uint diff = abs(x_diff) + abs(y_diff);
-
-		if (diff < rvf->best_diff || (diff == rvf->best_diff && v->index < rvf->best->index)) {
-			rvf->best = v;
-			rvf->best_diff = diff;
-		}
+	/* Not a close Road vehicle when it's not a road vehicle, in the depot, or ourself. */
+	if (v->type != VEH_ROAD || v->IsInDepot() || rvf->veh->First() != v->First()) return nullptr;
+
+	/* Not close when at a different height or when going in a different direction. */
+	if (abs(v->z_pos - rvf->veh->z_pos) >= 6 || v->direction != rvf->dir) return nullptr;
+
+	/* We 'return' the closest vehicle, in distance and then VehicleID as tie-breaker. */
+	uint diff = abs(x_diff) + abs(y_diff);
+	if (diff > rvf->best_diff || (diff == rvf->best_diff && v->index > rvf->best->index)) return nullptr;
```

Comment:
2nd `==` should be a `!=`. `||`  and `&&` need swapping.

Reply:
I don't think so. There isn't a second `==`, so might it be the first one?
If I swap `||` and `&&`, then this will only return true when `diff > rvf->best_diff`, as `diff != rvf->best_diff` would always be true because `diff` is  bigger. If the `==` should remain, then the equals test is pointless as it would always be false because `diff` is bigger.

This line should now read: 'If the vehicle is further or (the distance is the same and vehicle id is later), then don't bother with the rest as we won't be the best'.

Reply:
Ok, I just assumed it was all de-morgan.